File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        char_map = {}
        start = 0  # Start of the sliding window
        max_len = 0  # To store the maximum length found
        for i in range(len(s)):
            char = s[i]
            # If the character is in the window, move start to the right of its last position
            if char in char_map and char_map[char] >= start:
                start = char_map[char] + 1
            
            # Update the last position of the character
            char_map[char] = i
            
            # Calculate the window size and update max_len if necessary
            max_len = max(max_len, i - start + 1)
        
        return max_len
        

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int hash[256] = { -1 };  // Initialize all positions to -1
        fill(hash, hash + 256, -1); // Explicitly set all to -1 to indicate no character is seen yet

        int start = 0; // Start of the sliding window
        int maxLen = 0; // To store the maximum length found

        for (int i = 0; i < s.length(); ++i) {
            char ch = s[i];

            // If the character is already in the current window, move `start` forward
            if (hash[ch] >= start) {
                start = hash[ch] + 1;
            }

            // Update the last seen index of the character
            hash[ch] = i;

            // Update the maximum length of the substring
            maxLen = max(maxLen, i - start + 1);
        }

        return maxLen;
    }
};
